File: solutions/cpp/Update_BST.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

using std::cout;
using std::endl;
using std::unique_ptr;

// @include
class BinarySearchTree {
 public:
  virtual ~BinarySearchTree() { Clear(); }

  bool Empty() const { return !root_.get(); }

  void Clear() { Clear(&root_); }

  bool Insert(int key) {
    if (Empty()) {
      root_ = unique_ptr<TreeNode>(new TreeNode{key, nullptr, nullptr});
    } else {
      TreeNode* curr = root_.get(), *par;
      while (curr) {
        par = curr;
        if (key == curr->data) {
          return false;  // no insertion for duplicate key.
        } else if (key < curr->data) {
          curr = curr->left.get();
        } else {  // key > curr->data.
          curr = curr->right.get();
        }
      }

      // Inserts key according to key and par.
      if (key < par->data) {
        par->left.reset(new TreeNode{key});
      } else {
        par->right.reset(new TreeNode{key});
      }
    }
    return true;
  }

  bool Erase(int key) {
    // Finds the node with key.
    TreeNode* curr = root_.get(), *par = nullptr;
    while (curr && curr->data != key) {
      par = curr;
      curr = key < curr->data ? curr->left.get() : curr->right.get();
    }

    // No node with key in this binary tree.
    if (!curr) {
      return false;
    }

    if (curr->right) {
      // Finds the minimum of the right subtree.
      TreeNode* r_curr = curr->right.get(), *r_par = curr;
      while (r_curr->left) {
        r_par = r_curr;
        r_curr = r_curr->left.get();
      }
      // Moves links to erase the node.
      r_curr->left.reset(curr->left.release());
      TreeNode* r_curr_right = r_curr->right.release();
      if (curr->right.get() != r_curr) {
        r_curr->right.reset(curr->right.release());
      }
      if (r_par->left.get() == r_curr) {
        r_curr = r_par->left.release();
        r_par->left.reset(r_curr_right);
      } else {  // r_par->right.get() == r_curr.
        r_curr = r_par->right.release();
        r_par->right.reset(r_curr_right);
      }
      ReplaceParentChildLink(par, curr, r_curr);

      // Updates root_ link if needed.
      if (root_.get() == curr) {
        root_.reset(r_curr);
      }
    } else {
      // Updates root_ link if needed.
      if (root_.get() == curr) {
        root_.reset(curr->left.release());
      }
      ReplaceParentChildLink(par, curr, curr->left.get());
    }
    return true;
  }
  // @exclude
  int GetRootVal() const { return root_->data; }
  // @include

 private:
  struct TreeNode {
    int data;
    unique_ptr<TreeNode> left, right;
  };

  void Clear(unique_ptr<TreeNode>* node) {
    if (node->get()) {
      if ((*node)->left.get()) {
        Clear(&((*node)->left));
      }
      if ((*node)->right.get()) {
        Clear(&((*node)->right));
      }
      node->reset(nullptr);
    }
  }

  // Replaces the link between par and child by new_link.
  void ReplaceParentChildLink(TreeNode* par,
                              TreeNode* child,
                              TreeNode* new_link) {
    if (!par) {
      return;
    }

    if (par->left.get() == child) {
      par->left.reset(new_link);
    } else {  // par->right.get() == child.
      par->right.reset(new_link);
    }
  }

  unique_ptr<TreeNode> root_ = nullptr;
};
// @exclude

int main(int argc, char* argv[]) {
  BinarySearchTree BST;
  assert(BST.Empty() == true);
  assert(BST.Insert(4) == true);
  assert(BST.Insert(5) == true);
  assert(BST.Insert(2) == true);
  assert(BST.Insert(3) == true);
  assert(BST.Insert(1) == true);
  assert(BST.Empty() == false);
  assert(BST.Erase(0) == false);
  assert(BST.Erase(2) == true);
  assert(BST.Erase(2) == false);
  assert(BST.Insert(4) == false);
  // should output 4
  assert(BST.GetRootVal() == 4);
  cout << BST.GetRootVal() << endl;
  assert(BST.Erase(4) == true);
  // should output 5
  assert(BST.GetRootVal() == 5);
  cout << BST.GetRootVal() << endl;
  assert(BST.Erase(5) == true);
  // should output 3
  assert(BST.GetRootVal() == 3);
  cout << BST.GetRootVal() << endl;
  assert(BST.Erase(3) == true);
  // should output 1
  assert(BST.GetRootVal() == 1);
  cout << BST.GetRootVal() << endl;
  assert(BST.Erase(1) == true);
  assert(BST.Empty() == true);
  return 0;
}

File: solutions/python/update_bst.py
from binary_tree_prototype import BinaryTreeNode


# @include
class BinarySearchTree:

    def __init__(self):
        self._root = None

    # @exclude
    def empty(self):
        return self._root is None

    # @include
    def insert(self, key):
        if self.empty():
            self._root = BinaryTreeNode(key)
        else:
            parent = None
            curr = self._root
            while curr:
                parent = curr
                if key == curr.data:
                    # key already present, no duplicates to be added.
                    return False
                elif key < curr.data:
                    curr = curr.left
                else:  # key > curr.data.
                    curr = curr.right

            # Inserts key according to key and parent.
            if key < parent.data:
                parent.left = BinaryTreeNode(key)
            else:
                parent.right = BinaryTreeNode(key)
        return True

    def delete(self, key):
        # Find the node with key.
        curr = self._root
        parent = None
        while curr and curr.data != key:
            parent = curr
            curr = curr.left if key < curr.data else curr.right

        if not curr:
            # There's no node with key in this tree.
            return False
        key_node = curr

        if key_node.right:
            # Finds the minimum of the right subtree.
            r_key_node = key_node.right
            r_parent = key_node
            while r_key_node.left:
                r_parent = r_key_node
                r_key_node = r_key_node.left
            # Moves links to erase the node.
            r_key_node.left = key_node.left
            key_node.left = None
            if r_parent.left is r_key_node:
                r_parent.left = r_key_node.right
                r_key_node.right = key_node.right
            key_node.right = None
            self._replace_parent_child_link(parent, key_node, r_key_node)

            # Updates _root link if needed.
            if self._root is key_node:
                self._root = r_key_node
        else:
            # Updates _root link if needed.
            if self._root is key_node:
                self._root = key_node.left
            self._replace_parent_child_link(parent, key_node, key_node.left)
            key_node.left = None
        return True

    # @exclude

    def get_root_val(self):
        return self._root.data

    # @include

    # Replaces the link between parent and child by new_link.
    @staticmethod
    def _replace_parent_child_link(parent, child, new_link):
        if not parent:
            return

        if parent.left is child:
            parent.left = new_link
        else:  # parent.right is child.
            parent.right = new_link


# @exclude


def main():
    bst = BinarySearchTree()
    assert bst.empty() is True
    assert bst.insert(7) is True
    assert bst.insert(8) is True
    assert bst.insert(9) is True
    assert bst.insert(4) is True
    assert bst.insert(3) is True
    assert bst.empty() is False
    assert bst.insert(2) is True
    assert bst.insert(5) is True
    assert bst.delete(7) is True
    assert bst.delete(9) is True
    # should output 8
    assert bst.get_root_val() == 8
    print(bst.get_root_val())
    assert bst.delete(4) is True
    # should output 8
    assert bst.get_root_val() == 8
    print(bst.get_root_val())
    assert bst.delete(8) is True
    # should output 5
    assert bst.get_root_val() == 5
    print(bst.get_root_val())
    assert bst.delete(5) is True
    assert bst.delete(3) is True
    # should output 2
    assert bst.get_root_val() == 2
    print(bst.get_root_val())
    assert bst.delete(2) is True
    assert bst.delete(1) is False
    assert bst.empty() is True
    bst.insert(7)
    assert bst.get_root_val() == 7
    bst.insert(9)
    assert bst.get_root_val() == 7
    bst.delete(7)
    assert bst.get_root_val() == 9


if __name__ == '__main__':
    main()
